### ISSUE TITLE
Fail to authenticate AKV in Azure VM

### ISSUE THREAD

**Issue by shizhMSFT** [Azure/notation-azure-kv#29]
On Azure VM with MI disabled, the plugin failed to authenticate AKV as the client tries to obtain a MI token with the following message:

```
Error: describe-key command failed: azure-kv: : azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://<sanitized_kv_name>.vault.azure.net/secrets/wabbit-networks-io/4df1ba8a0f374f8c8507d1bdfdd6c13e?api-version=7.1: StatusCode=400 -- Original Error: adal: Refresh request failed. Status Code = '400'. Response body: {"error":"invalid_request","error_description":"Identity not found"} Endpoint http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https%3A%2F%2Fvault.azure.net
2022/08/18 14:29:32 describe-key command failed: azure-kv: : azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://<sanitized_kv_name>.vault.azure.net/secrets/wabbit-networks-io/4df1ba8a0f374f8c8507d1bdfdd6c13e?api-version=7.1: StatusCode=400 -- Original Error: adal: Refresh request failed. Status Code = '400'. Response body: {"error":"invalid_request","error_description":"Identity not found"} Endpoint http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https%3A%2F%2Fvault.azure.net
```

The plugin should fallback to case 5 **Azure CLI 2.0**. However, it does not since `err` from line 25 is `nil` and makes line 26 false.

https://github.com/Azure/notation-azure-kv/blob/b8f13d7ffe1b710a5dd6d016b375e282c5fabeaa/internal/cloud/keyvault.go#L18-L37

**Comment by FeynmanZhou** [comment]
I tested it by following [this guide](https://github.com/chloeyin/notation-azure-kv/blob/cose/docs/quick-start.md), but I got the error msg:
```
$ notation sign --key ${KEY_NAME} ${IMAGE}
Error: failed to sign digest. error : generate-signature command failed: azure-kv: ERROR: keyvault.BaseClient#Sign: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="Forbidden" Message="The user, group or application 'appid=04b07795-8ddb-461a-bbee-02f9e1bf7b46;oid=fa59d25d-1105-4674-b90f-e11f751a96b0;numgroups=696;iss=https://sts.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47/' does not have keys sign permission on key vault 'feynman-akv;location=eastus'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287" InnerError={"code":"ForbiddenByPolicy"}
2022/08/18 16:46:16 failed to sign digest. error : generate-signature command failed: azure-kv: ERROR: keyvault.BaseClient#Sign: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="Forbidden" Message="The user, group or application 'appid=04b07795-8ddb-461a-bbee-02f9e1bf7b46;oid=fa59d25d-1105-4674-b90f-e11f751a96b0;numgroups=696;iss=https://sts.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47/' does not have keys sign permission on key vault 'feynman-akv;location=eastus'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287" InnerError={"code":"ForbiddenByPolicy"}]
```

**Comment by shizhMSFT** [comment]
> I tested it by following [this guide](https://github.com/chloeyin/notation-azure-kv/blob/cose/docs/quick-start.md), but I got the error msg:
> 
> ```
> $ notation sign --key ${KEY_NAME} ${IMAGE}
> Error: failed to sign digest. error : generate-signature command failed: azure-kv: ERROR: keyvault.BaseClient#Sign: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="Forbidden" Message="The user, group or application 'appid=04b07795-8ddb-461a-bbee-02f9e1bf7b46;oid=fa59d25d-1105-4674-b90f-e11f751a96b0;numgroups=696;iss=https://sts.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47/' does not have keys sign permission on key vault 'feynman-akv;location=eastus'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287" InnerError={"code":"ForbiddenByPolicy"}
> 2022/08/18 16:46:16 failed to sign digest. error : generate-signature command failed: azure-kv: ERROR: keyvault.BaseClient#Sign: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="Forbidden" Message="The user, group or application 'appid=04b07795-8ddb-461a-bbee-02f9e1bf7b46;oid=fa59d25d-1105-4674-b90f-e11f751a96b0;numgroups=696;iss=https://sts.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47/' does not have keys sign permission on key vault 'feynman-akv;location=eastus'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287" InnerError={"code":"ForbiddenByPolicy"}]
> ```

That's a permission issue. You need to grant `sign` permission for that key in AKV.

**Comment by FeynmanZhou** [comment]
> > I tested it by following [this guide](https://github.com/chloeyin/notation-azure-kv/blob/cose/docs/quick-start.md), but I got the error msg:
> > ```
> > $ notation sign --key ${KEY_NAME} ${IMAGE}
> > Error: failed to sign digest. error : generate-signature command failed: azure-kv: ERROR: keyvault.BaseClient#Sign: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="Forbidden" Message="The user, group or application 'appid=04b07795-8ddb-461a-bbee-02f9e1bf7b46;oid=fa59d25d-1105-4674-b90f-e11f751a96b0;numgroups=696;iss=https://sts.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47/' does not have keys sign permission on key vault 'feynman-akv;location=eastus'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287" InnerError={"code":"ForbiddenByPolicy"}
> > 2022/08/18 16:46:16 failed to sign digest. error : generate-signature command failed: azure-kv: ERROR: keyvault.BaseClient#Sign: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="Forbidden" Message="The user, group or application 'appid=04b07795-8ddb-461a-bbee-02f9e1bf7b46;oid=fa59d25d-1105-4674-b90f-e11f751a96b0;numgroups=696;iss=https://sts.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47/' does not have keys sign permission on key vault 'feynman-akv;location=eastus'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287" InnerError={"code":"ForbiddenByPolicy"}]
> > ```
> 
> That's a permission issue. You need to grant `sign` permission for that key in AKV.

Granted and resolved. But I still got 401 error returns as follows:

```
Error: HEAD "https://feynmanacr.azurecr.io/v2/net-monitor/manifests/v1": GET "https://feynmanacr.azurecr.io/oauth2/token?scope=repository%3Anet-monitor%3Apull&service=feynmanacr.azurecr.io": unexpected status code 401: unauthorized: authentication required, visit https://aka.ms/acr/authorization for more information.
2022/08/19 07:40:15 HEAD "https://feynmanacr.azurecr.io/v2/net-monitor/manifests/v1": GET "https://feynmanacr.azurecr.io/oauth2/token?scope=repository%3Anet-monitor%3Apull&service=feynmanacr.azurecr.io": unexpected status code 401: unauthorized: authentication required, visit https://aka.ms/acr/authorization for more information.
```

**Comment by FeynmanZhou** [comment]
> > > I tested it by following [this guide](https://github.com/chloeyin/notation-azure-kv/blob/cose/docs/quick-start.md), but I got the error msg:
> > > ```
> > > $ notation sign --key ${KEY_NAME} ${IMAGE}
> > > Error: failed to sign digest. error : generate-signature command failed: azure-kv: ERROR: keyvault.BaseClient#Sign: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="Forbidden" Message="The user, group or application 'appid=04b07795-8ddb-461a-bbee-02f9e1bf7b46;oid=fa59d25d-1105-4674-b90f-e11f751a96b0;numgroups=696;iss=https://sts.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47/' does not have keys sign permission on key vault 'feynman-akv;location=eastus'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287" InnerError={"code":"ForbiddenByPolicy"}
> > > 2022/08/18 16:46:16 failed to sign digest. error : generate-signature command failed: azure-kv: ERROR: keyvault.BaseClient#Sign: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="Forbidden" Message="The user, group or application 'appid=04b07795-8ddb-461a-bbee-02f9e1bf7b46;oid=fa59d25d-1105-4674-b90f-e11f751a96b0;numgroups=696;iss=https://sts.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47/' does not have keys sign permission on key vault 'feynman-akv;location=eastus'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287" InnerError={"code":"ForbiddenByPolicy"}]
> > > ```
> > 
> > 
> > That's a permission issue. You need to grant `sign` permission for that key in AKV.
> 
> Granted and resolved. But I still got 401 error returns as follows:
> 
> ```
> Error: HEAD "https://feynmanacr.azurecr.io/v2/net-monitor/manifests/v1": GET "https://feynmanacr.azurecr.io/oauth2/token?scope=repository%3Anet-monitor%3Apull&service=feynmanacr.azurecr.io": unexpected status code 401: unauthorized: authentication required, visit https://aka.ms/acr/authorization for more information.
> 2022/08/19 07:40:15 HEAD "https://feynmanacr.azurecr.io/v2/net-monitor/manifests/v1": GET "https://feynmanacr.azurecr.io/oauth2/token?scope=repository%3Anet-monitor%3Apull&service=feynmanacr.azurecr.io": unexpected status code 401: unauthorized: authentication required, visit https://aka.ms/acr/authorization for more information.
> ```

Resolved 401 error. It requires configuring authentication to my ACR repo.

**Comment by chloeyin** [comment]
I will take a look at this.

**Comment by FeynmanZhou** [comment]
Thanks @chloeyin . 

Let me summarize the to-do items to resolve this issue and ensure Notation CLI v0.10.0-alpha.3 works fine with this plugin:

- [ ] Fix the authentication issue of the Notation Azure KV plugin and provide a user-friendly method to configure the location of the client credential (Considering passing environmental variables)  
- [ ] Publish a new release (v0.4.0-alpha.1) for the Notation Azure KV plugin
- [ ] Functional testing for this new release
- [ ] Update the Azure Doc: [Build, sign, and verify container images using Notary and Azure Key Vault (Preview)](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-tutorial-sign-build-push)

cc @dtzar @SteveLasker could you pls help to assign this issue to @chloeyin and me?

**Comment by FeynmanZhou** [comment]
There will be a certificate chain update in Notation CLI v0.10.0-alpha.4, we can first introduce the fix of the authentication issue in Azure VM https://github.com/Azure/notation-azure-kv/pull/30 and release a fix pack `v0.3.1-alpha.1` for the Notation Azure KV plugin. 

I tested the latest build of the AKV plugin with Notation CLI v0.9.0-alpha.1, it works fine. Let's prepare for releasing `v0.3.1-alpha.1`.

**Comment by shizhMSFT** [comment]
Closed by #30
